src/ocean/util/container/btree/BTreeMap.d

```diff
@@ -148,9 +148,40 @@ struct BTreeMap(TreeKeyType, TreeValueType, int tree_degree)
 
     public bool insert (KeyType key, ValueType value)
     {
-        return this.impl.insert(key, value);
+        bool added;
+        this.impl.insert(key, value, added);
+        return added;
     }
 
+    /***************************************************************************
+
+        Inserts the (key, value) in the tree. This is passing the copy of the
+        value into the tree, so it's not necessary to manually create copy of
+        it.  Note that for the reference types, this will just copy the
+        reference.
+
+        Params:
+            key = key to insert
+            value = value associated to the key to insert.
+            added = indicator if the value has been added, or the value
+                associated with the key was already in the map
+
+        Returns:
+            Pointer to the element associated with the given key (either existing
+            or new value).
+
+        Complexity:
+            CPU: O(degree * log_degree(n))
+            Memory:O(log_degree(n))
+
+    ***************************************************************************/
+
+    public ValueType* insert (KeyType key, ValueType value, out bool added)
```

Comment:
Meh, seeing more of the `out bool something` pattern I ask rest of core team to reconsider addition of `Optional` type into ocean :)

Reply:
I think that's not a bad idea at all.

Reply:
If you make a PR to get it included in ocean, I'll go and base this PR on top of that ;-)

Reply:
I'm now looking again into this and I think there's no need for `Optional` here :-(. Trick is that this method always returns a pointer, either to the old value or the new inserted value, passing the `bool` to the user to say if it added or removed the new element (it's akin to `getOrCreate` in Ocean's caches, really). Do you have proposal for the better API?

Reply:
Ha. Yeah, this function's API was misread. So you could add `out (p) { assert(p !is null); }`.

Reply:
Yes, will add, along with the unittests.

Reply:
:(